### DISCUSSION TITLE
Density parameter missing in material models?

### DISCUSSION THREAD

**Discussion by mrgprasad** [thelfer/tfel#373] [General]
I am comparing codeaster's Signorini model with MFront.
The [Signorini.mfront ](https://tfel.sourceforge.net/gallery/hyperelasticity/Signorini.mfront)doesn't take Density (Rho) as a input parameter. I understand that the constitutive equation don't need density as input. But the density influences the stiffness and damping in a material, hence it has to be defined in the simulations somehow when using the MFront models. 

Here's the code aster definition of hyperelastic material
```
ELAS_HYPER = _F(
                  ♦ C10 = c10,                                [R]
                  ◊ C01 = / c01,                              [R]
                          / 0.0,                              [DEFAULT]
                  ◊ C20 = / c20,                              [R]
                          / 0.0,                              [DEFAULT]
                  ◊ RHO = / rho,                              [R]
                          / 0.0,                              [DEFAULT]
                  ◊ NU = poisson,                             [R]
                  ◊ K = bulk                                  [R]
                 )
```

In the codeaster .comm file, hyperelastic material is defined as:
```
C01 = 2.3456;
C10 = 0.709;
C20 = 0.0;
NU  = 0.499
K   = (6*(C10+C01))/(3*(1-2*NU))
RUBBER1=DEFI_MATERIAU(ELAS_HYPER=_F(C10=C10,
                                   C01=C01,
                                   C20=C20,
                                   K=K,
                                   RHO=1000.0),);
```
However while using MFront, the material will be defined as:
```
RUBBER1=DEFI_MATERIAU(MFRONT=_F( LISTE_COEF=(C01,C10,C20,K),),);
```
@thelfer My question is then how does one define density in codeaster when using MFront material models.

**Comment by thelfer** [comment]
@mrgprasad The mass density can be defined in the material properties. Then you have to explicitly implement the dependencies of the elastic properties to the mass density in `MFront`.

You probably also define the mass density for dynamic analysis in `code_aster` but I have no experience in that matter.

**Reply by mrgprasad** [reply]
@thelfer To account for the effect of the mass density on the elastic properties of the material we could scale them with the mass density like this right?

```
@MaterialProperty rho; // scale factor for mass density (= 1.0 for example)
@MaterialProperty stress K;
@MaterialProperty stress mu;
@MaterialProperty real alpha;

K = K*rho
mu = mu*rho
alpha = alpha*rho
```

**Reply by thelfer** [reply]
@mrgprasad You shall never modify material properties, that's a bad practice. In future versions of `MFront`, they will probably be declared immutable (`const` in C++). I would do:

```cxx
@MaterialProperty massdensity rho; // scale factor for mass density (= 1.0 for example)
//! reference value of the bulk modulus, i.e. the value of the bulk modulus for....
@MaterialProperty stress K_ref;
K_ref.setEntryName("ReferenceValueForTheBulkModulus");
@MaterialProperty stress mu_ref;
@MaterialProperty real alpha_ref;

@LocalVariables stress K, mu;
@LocalVariables real alpha;

@InitLocalVariables{
K = K_ref*rho
mu = mu_ref*rho
alpha = alpha_ref*rho
}
```
